core/src/com/unciv/logic/BarbarianManager.kt

```diff
@@ -0,0 +1,269 @@
+            .flatMap { it.getCapital().getCenterTile().getTilesInDistance(4) }.toSet()
+        val tooCloseToCamps = camps
+            .flatMap { tileMap[it.key].getTilesInDistance(7) }.toSet()
+
+        val viableTiles = fogTiles.filter {
+            !it.isImpassible()
+                    && it.resource == null
+                    && it.terrainFeatures.none { feature -> gameInfo.ruleSet.terrains[feature]!!.hasUnique("Only [] improvements may be built on this tile") }
+                    && it.neighbors.any { neighbor -> neighbor.isLand }
+                    && it !in tooCloseToCapitals
+                    && it !in tooCloseToCamps
+        }.toMutableList()
+
+        var tile: TileInfo?
+        var addedCamps = 0
+        var biasCoast = Random().nextInt(6) == 0
+
+        // Add the camps
+        while (addedCamps < campsToAdd) {
+            if (viableTiles.isEmpty())
+                break
+
+            // If we're biasing for coast, get a coast tile if possible
+            if (biasCoast) {
+                tile = viableTiles.filter { it.isCoastalTile() }.randomOrNull()
+                if (tile == null)
+                    tile = viableTiles.random()
+            } else
+                tile = viableTiles.random()
+            
+            tile.improvement = Constants.barbarianEncampment
+            val newCamp = Encampment()
+            newCamp.position = tile.position
+            newCamp.gameInfo = gameInfo
+            camps[newCamp.position] = newCamp
+            notifyCivsOfBarbarianEncampment(tile)
+            addedCamps++
+
+            // Still more camps to add?
+            if (addedCamps < campsToAdd) {
+                // Remove some newly non-viable tiles
+                viableTiles.removeAll( tile.getTilesInDistance(7) )
+                // Reroll bias
+                biasCoast = Random().nextInt(6) == 0
+            }
+        }
+    }
+
+    /**
+     * [CivilizationInfo.addNotification][Add a notification] to every civilization that have
+     * adopted Honor policy and have explored the [tile] where the Barbarian Encampment has spawned.
+     */
+    private fun notifyCivsOfBarbarianEncampment(tile: TileInfo) {
+        gameInfo.civilizations.filter {
+            it.hasUnique("Notified of new Barbarian encampments")
+                    && it.exploredTiles.contains(tile.position)
+        }
+            .forEach { it.addNotification("A new barbarian encampment has spawned!", tile.position, NotificationIcon.War) }
+    }
+}
+
+class Encampment {
+    var countdown = 0
+    var spawnedUnits = -1
+    lateinit var position: Vector2
+
+    @Transient
+    lateinit var gameInfo: GameInfo
+
+    fun clone(): Encampment {
+        val toReturn = Encampment()
+        toReturn.position = position
+        toReturn.countdown = countdown
+        toReturn.spawnedUnits = spawnedUnits
+        return toReturn
+    }
+
+    fun update() {
+        if (countdown <= 0) {
```

Comment:
Switch around for if ... Else if ... For less indentation and easier readability